From cf3fb498b2789ab57f51349d41658ffd9b3a8bc5 Mon Sep 17 00:00:00 2001
From: Pieter De Gendt <pieter.degendt@basalte.be>
Date: Thu, 19 Dec 2024 07:31:22 +0100
Subject: [PATCH] scripts: west_commands: Support out-of-tree runners

Add runners entry to the module schema and import the file specified.

Every class that inherits from ZephyrBinaryRunner will be discovered and
added to runners list.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>
---
 scripts/west_commands/run_common.py | 20 +++++++++++++++++++-
 scripts/zephyr_module.py            |  9 +++++++++
 2 files changed, 28 insertions(+), 1 deletion(-)

diff --git a/scripts/west_commands/run_common.py b/scripts/west_commands/run_common.py
index 012f27359d6..b807dd03bd3 100644
--- a/scripts/west_commands/run_common.py
+++ b/scripts/west_commands/run_common.py
@@ -6,6 +6,7 @@
 '''Common code used by commands which execute runners.
 '''
 
+import importlib.util
 import re
 import argparse
 import logging
@@ -28,7 +29,9 @@ from runners.core import FileType
 from runners.core import BuildConfiguration
 import yaml
 
-from zephyr_ext_common import ZEPHYR_SCRIPTS
+from zephyr_ext_common import ZEPHYR_BASE, ZEPHYR_SCRIPTS
+sys.path.insert(0, str(ZEPHYR_SCRIPTS))
+import zephyr_module
 
 # Runners depend on edtlib. Make sure the copy in the tree is
 # available to them before trying to import any.
@@ -107,6 +110,13 @@ class SocBoardFilesProcessing:
     priority: int = IGNORED_RUN_ONCE_PRIORITY
     yaml: object = None
 
+def import_from_path(module_name, file_path):
+    spec = importlib.util.spec_from_file_location(module_name, file_path)
+    module = importlib.util.module_from_spec(spec)
+    sys.modules[module_name] = module
+    spec.loader.exec_module(module)
+    return module
+
 def command_verb(command):
     return "flash" if command.name == "flash" else "debug"
 
@@ -197,6 +207,14 @@ def do_run_common(command, user_args, user_runner_args, domain_file=None):
         dump_context(command, user_args, user_runner_args)
         return
 
+    # Import external module runners
+    for module in zephyr_module.parse_modules(ZEPHYR_BASE, command.manifest):
+        runners_ext = module.meta.get("runners", [])
+        for runner in runners_ext:
+            import_from_path(
+                module.meta.get("name", "runners_ext"), Path(module.project) / runner["file"]
+            )
+
     build_dir = get_build_dir(user_args)
     if not user_args.skip_rebuild:
         rebuild(command, build_dir, user_args)
diff --git a/scripts/zephyr_module.py b/scripts/zephyr_module.py
index c6043154bdc..9b8a52fb629 100755
--- a/scripts/zephyr_module.py
+++ b/scripts/zephyr_module.py
@@ -161,6 +161,15 @@ mapping:
          type: seq
          sequence:
             - type: str
+  runners:
+    required: false
+    type: seq
+    sequence:
+      - type: map
+        mapping:
+          file:
+            required: true
+            type: str
 '''
 
 MODULE_YML_PATH = PurePath('zephyr/module.yml')
-- 
2.48.1

